src/arm-codegen.c

```diff
@@ -445,3 +445,358 @@ void code_generate()
+                case OP_lshift:
+                case OP_rshift:
+                case OP_bit_and:
+                case OP_bit_or:
+                case OP_bit_xor:
+                case OP_negate:
+                case OP_log_and:
+                case OP_bit_not:
+                    flatten_ir = add_ph2_ir(inst->op);
+                    memcpy(flatten_ir, inst, sizeof(ph2_ir_t));
+                    elf_offset += 4;
+                    break;
+                case OP_load_data_address:
+                    flatten_ir = add_ph2_ir(OP_load_data_address);
+                    memcpy(flatten_ir, inst, sizeof(ph2_ir_t));
+                    elf_offset += 8;
+                    break;
+                case OP_mod:
+                case OP_eq:
+                case OP_neq:
+                case OP_gt:
+                case OP_lt:
+                case OP_geq:
+                case OP_leq:
+                case OP_log_not:
+                case OP_log_or:
+                    flatten_ir = add_ph2_ir(inst->op);
+                    memcpy(flatten_ir, inst, sizeof(ph2_ir_t));
+                    elf_offset += 12;
+                    break;
+                case OP_branch:
+                    flatten_ir = add_ph2_ir(OP_branch);
+                    memcpy(flatten_ir, inst, sizeof(ph2_ir_t));
+                    elf_offset += 24;
+                    break;
+                case OP_return:
+                    flatten_ir = add_ph2_ir(OP_return);
+                    memcpy(flatten_ir, inst, sizeof(ph2_ir_t));
+                    flatten_ir->src1 = bb->belong_to->func->stack_size;
+                    elf_offset += 24;
+                    break;
+                default:
+                    printf("Unknown OP\n");
+                    abort();
+                }
+            }
+        }
+    }
+}
+
+void generate()
+{
+    int elf_data_start = elf_code_start + elf_offset;
+
+    emit(__movw(__AL, __r8, GLOBAL_FUNC.stack_size));
+    emit(__movt(__AL, __r8, GLOBAL_FUNC.stack_size));
+    emit(__sub_r(__AL, __sp, __sp, __r8));
+    emit(__mov_r(__AL, __r12, __sp));
+    emit(__bl(__AL, MAIN_BB->elf_offset - elf_code_idx));
+
+    emit(__movw(__AL, __r8, GLOBAL_FUNC.stack_size));
+    emit(__movt(__AL, __r8, GLOBAL_FUNC.stack_size));
+    emit(__add_r(__AL, __sp, __sp, __r8));
+    emit(__mov_r(__AL, __r0, __r0));
+    emit(__mov_i(__AL, __r7, 1));
+    emit(__svc());
+
+    emit(__mov_r(__AL, __r7, __r0));
+    emit(__mov_r(__AL, __r0, __r1));
+    emit(__mov_r(__AL, __r1, __r2));
+    emit(__mov_r(__AL, __r2, __r3));
+    emit(__mov_r(__AL, __r3, __r4));
+    emit(__mov_r(__AL, __r4, __r5));
+    emit(__mov_r(__AL, __r5, __r6));
+    emit(__svc());
+    emit(__mov_r(__AL, __pc, __lr));
+
+    int i;
+    for (i = 0; i < ph2_ir_idx; i++) {
+        ph2_ir_t *ph2_ir = &PH2_IR[i];
+        int rd = ph2_ir->dest;
+        int rn = ph2_ir->src0;
+        int rm = ph2_ir->src1;
+        int ofs;
+
+        switch (ph2_ir->op) {
```

Comment:
Can you move some parts of this function out? It is lengthy.

Reply:
Using the compiler directive could extract this part from both ARM and RISC-V backend, but the readability might be worse. Is it recommended?

Reply:
No, simply use function calls to split the code.